pennylane/labs/dla/dense_util.py

```diff
@@ -0,0 +1,456 @@
+        "or not all operators are of the same type."
+    )
+
+
+def _gram_ps(basis: Iterable[PauliSentence]):
+    gram = np.zeros((len(basis), len(basis)))
+    for (i, b_i), (j, b_j) in combinations_with_replacement(enumerate(basis), r=2):
+        gram[i, j] = gram[j, i] = (b_i @ b_j).trace()
+    return gram
+
+
+def _op_to_adjvec_ps(ops: PauliSentence, basis: PauliSentence, is_orthogonal: bool = True):
+    """Pauli sentence branch of ``op_to_adjvec``."""
+    if not all(isinstance(op, PauliSentence) for op in ops):
+        ops = [op.pauli_rep for op in ops]
+
+    res = []
+    if is_orthogonal:
+        norms_squared = [(basis_i @ basis_i).trace() for basis_i in basis]
+    else:
+        # Fake the norm correction if we anyways will apply the inverse Gram matrix later
+        norms_squared = np.ones(len(basis))
+        gram = _gram_ps(basis)
+        inv_gram = np.linalg.pinv(sqrtm(gram))
+
+    for op in ops:
+        rep = np.zeros((len(basis),))
+        for i, basis_i in enumerate(basis):
+            # v = ∑ (v · e_j / ||e_j||^2) * e_j
+            rep[i] = (basis_i @ op).trace() / norms_squared[i]
+
+        res.append(rep)
+    res = np.array(res)
+    if not is_orthogonal:
+        res = np.einsum("ij,kj->ki", inv_gram, res)
+
+    return res
+
+
+def op_to_adjvec(
+    ops: Iterable[Union[PauliSentence, Operator, np.ndarray]],
+    basis: Union[PauliSentence, Operator, np.ndarray],
+    is_orthogonal: bool = True,
+):
+    r"""Decompose a batch of operators onto a given operator basis.
+
+    The adjoint vector representation are the coefficients :math:`c_j` in a given operator
+    basis of the operator :math:`\hat{b}_j` such that the input operator can be written as
+    :math:`\hat{O} = \sum_j c_j \hat{b}_j`.
+
+    .. seealso:: :func:`~adjvec_to_op`
+
+    Args:
+        ops (Iterable[Union[PauliSentence, Operator, np.ndarray]]): List of operators to decompose
+        basis (Iterable[Union[PauliSentence, Operator, np.ndarray]]): Operator basis
+        is_orthogonal (bool): Whether the basis is orthogonal with respect to the trace inner
+            product. Defaults to ``True``, which allows to skip some computations.
+
+    Returns:
+        np.ndarray: The batch of coefficient vectors of the operators ``ops`` expressed in
+        ``basis``. The shape is ``(len(ops), len(basis)``.
+
+    The format of the resulting operators is determined by the ``type`` in ``basis``.
+    If ``is_orthogonal=True`` (the default), only normalization is taken into account
+    in the projection. For ``is_orthogonal=False``, orthogonalization also is considered.
+
+    **Example**
+
+    The basis can be numerical or operators.
+
+    >>> from pennylane.labs.dla import op_to_adjvec
+    >>> op = X(0) + 0.5 * Y(0)
+    >>> basis = [qml.X(0), qml.Y(0), qml.Z(0)]
+    >>> op_to_adjvec([op], basis)
+    array([[1. , 0.5, 0. ]])
+    >>> op_to_adjvec([op], [op.matrix() for op in basis])
+    array([[1. +0.j, 0.5+0.j, 0. +0.j]])
```

Comment:
Do we actually still produce complex valued outputs?
Do we want that?

Reply:
I think here yes because of the matrix reps in the basis

Reply:
no you're right, just hadnt updated my branch 👍

Reply:
fixed ✅